
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>GenBankParser &#8212; GenBankParser 0.1.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Welcome to GenBankParser’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="genbankparser">
<h1>GenBankParser<a class="headerlink" href="#genbankparser" title="Permalink to this headline">¶</a></h1>
<p><strong><em>Unofficial</em></strong> parser for ncbi GenBank data in the <em>GenBank flatfile format</em>.</p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p><em>I recommend using a <a class="reference external" href="https://pypi.org/project/virtualenv/">virtualenv</a>!</em></p>
<p>The packages can be pip-installed</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install git+git://github.com/j-i-l/GenBankParser.git@v0.1.1-alpha
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">v0.1.1-alpha</span></code> is the last version at the moment of writing these instructions.
Check the <a class="reference external" href="https://github.com/j-i-l/GenBankParser/releases">releases</a> section for newer versions.</p>
</div>
<div class="section" id="package-info">
<h1>Package info<a class="headerlink" href="#package-info" title="Permalink to this headline">¶</a></h1>
<p><strong>Contributions are welcome!</strong></p>
<p>This packages is not actively maintained.</p>
<div class="section" id="python-compatibility">
<h2>Python compatibility<a class="headerlink" href="#python-compatibility" title="Permalink to this headline">¶</a></h2>
<p>Supported is any python version &gt;=2.7 including python 3.x.</p>
</div>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p><a class="reference external" href="https://docs.python.org/2/library/configparser.html">configparser</a></p>
<blockquote>
<div><p>pip install configparser</p>
</div></blockquote>
</li>
<li><p><a class="reference external" href="http://docs.python-requests.org/en/master/">requests</a></p>
<blockquote>
<div><p>pip install requests</p>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="accepted-content">
<h2>Accepted content<a class="headerlink" href="#accepted-content" title="Permalink to this headline">¶</a></h2>
<p>This GenBankParser aims to parse uncompressed GenBank files in the <em>GenBank
flatfile format</em>.</p>
<p>They are usually of a form similar to this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">LOCUS</span>       <span class="n">XXXX</span>             <span class="mi">11111111</span> <span class="n">bp</span>    <span class="n">DNA</span>     <span class="n">circular</span> <span class="n">BCT</span> <span class="mi">01</span><span class="o">-</span><span class="n">JAN</span><span class="o">-</span><span class="mi">2018</span>
<span class="n">DEFINITION</span>  <span class="n">Completely</span> <span class="n">made</span> <span class="n">up</span><span class="p">,</span> <span class="n">complete</span> <span class="n">genome</span><span class="o">.</span>
<span class="n">ACCESSION</span>   <span class="n">XXXX</span>
<span class="n">VERSION</span>     <span class="n">XXXX</span><span class="o">.</span><span class="mi">1</span>  <span class="n">GI</span><span class="p">:</span><span class="mi">1111111111</span>
<span class="n">DBLINK</span>      <span class="n">BioProject</span><span class="p">:</span> <span class="n">PRJNA111111</span>
            <span class="n">BioSample</span><span class="p">:</span> <span class="n">SAMN111111</span>
<span class="n">KEYWORDS</span>    <span class="o">.</span>
<span class="n">SOURCE</span>      <span class="n">Completely</span> <span class="n">made</span> <span class="n">up</span>
  <span class="n">ORGANISM</span>  <span class="n">Completely</span> <span class="n">made</span> <span class="n">up</span>
            <span class="n">Bacteria</span><span class="p">;</span> <span class="n">Proteobacteria</span><span class="p">;</span> <span class="n">Gammaproteobacteria</span><span class="p">;</span> <span class="n">Enterobacterales</span><span class="o">.</span>
<span class="o">...</span>
</pre></div>
</div>
<p>Accepted are either files with single genomes or genes like <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?tool=portal&amp;save=file&amp;log%24=seqview&amp;db=nuccore&amp;report=gbwithparts&amp;id=22222&amp;withparts=on">this file</a> or a complete sequence of genomes available from the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/">NIH genetic sequence database</a>.</p>
<p>If you want to process sequence of genomes downloaded from the ncbi GenBank ftp server (ftp://ftp.ncbi.nih.gov/genbank/), please make sure to first decompress the files before using the GenBankParser.</p>
<p>In addition to GenBank files the GenBankParser also accepts GenBank UIDs or chromosome Genbank identifiers.
GenBankParser then tries to fetch the entries directly from the ncbi database. For an example see the <a class="reference external" href="#fetch-from-ncbi">example</a> below.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="use-cases">
<h1>Use cases<a class="headerlink" href="#use-cases" title="Permalink to this headline">¶</a></h1>
<div class="section" id="simple-parsing">
<h2>Simple parsing<a class="headerlink" href="#simple-parsing" title="Permalink to this headline">¶</a></h2>
<div class="section" id="get-a-list-of-genomes">
<h3>Get a list of genomes<a class="headerlink" href="#get-a-list-of-genomes" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">gbparse</span> <span class="kn">import</span> <span class="n">Parser</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">Parser</span><span class="p">()</span>

<span class="n">genome_file</span> <span class="o">=</span> <span class="s1">&#39;/path/to/genome_file.txt&#39;</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genome_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fobj</span><span class="p">:</span>
    <span class="n">genomes</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fobj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="save-genomes-as-json-files-to-a-directory">
<h3>Save genomes as json files to a directory<a class="headerlink" href="#save-genomes-as-json-files-to-a-directory" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">gbparse</span> <span class="kn">import</span> <span class="n">Parser</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">Parser</span><span class="p">()</span>

<span class="n">genome_file</span> <span class="o">=</span> <span class="s1">&#39;/path/to/genome_file.txt&#39;</span>
<span class="n">genomes_save_path</span> <span class="o">=</span> <span class="s1">&#39;/path/to/genomes/&#39;</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genome_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fobj</span><span class="p">:</span>
    <span class="n">genomes</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fobj</span><span class="p">,</span> <span class="n">genomes_save_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="processing">
<h2>Processing<a class="headerlink" href="#processing" title="Permalink to this headline">¶</a></h2>
<div class="section" id="retrieve-set-of-all-present-genes-in-genomes">
<h3>retrieve set of all present genes in genomes<a class="headerlink" href="#retrieve-set-of-all-present-genes-in-genomes" title="Permalink to this headline">¶</a></h3>
<p>You might pass a callable to the parser method. The callable needs to accept
a genome (a dictionary) as first argument but can de arbitrary otherwise.
Additional arguments can directly be passed to the parser method.</p>
<p>A simple use-case of a callable would be a method extracting certain
information from each parsed genome, like the set of present genes:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">gbparse</span> <span class="kn">import</span> <span class="n">Parser</span>

<span class="c1"># define a callable that retrieves all genes from a genomes</span>
<span class="k">def</span> <span class="nf">get_genes</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">present_genomes</span><span class="p">):</span>
    <span class="n">present_genomes</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
    	<span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span>
		<span class="n">gene</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>
		<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">[</span><span class="s1">&#39;content&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="p">{})</span>
	    <span class="p">))</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="bp">None</span>
    
<span class="n">p</span> <span class="o">=</span> <span class="n">Parser</span><span class="p">()</span>
<span class="c1"># define result variable</span>
<span class="n">list_of_present_genes</span> <span class="o">=</span> <span class="p">[]</span>

<span class="n">genome_file</span> <span class="o">=</span> <span class="s1">&#39;/path/to/genome_file.txt&#39;</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genome_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fobj</span><span class="p">:</span>
    <span class="n">p</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fobj</span><span class="p">,</span> <span class="n">fct</span><span class="o">=</span><span class="n">get_genes</span><span class="p">,</span> <span class="n">present_genomes</span><span class="o">=</span><span class="n">list_of_present_genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="fetch-from-ncbi">
<h2>Fetch from ncbi<a class="headerlink" href="#fetch-from-ncbi" title="Permalink to this headline">¶</a></h2>
<p>Say we want the get the first 10 GenBank files that are returned when searching for ‘hiv’ on the Pubmed database.
Using the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25500/">ncbi entrez eutils</a> tool the query to retrieve UID’s of these entries might look like this:</p>
<blockquote>
<div><p>https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=hiv&amp;retstart=0&amp;retmax=10&amp;rettype=text&amp;tool=biomed3&amp;format=json</p>
</div></blockquote>
<p>Here is how this can all be done in python:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">from</span> <span class="nn">gbparse</span> <span class="kn">import</span> <span class="n">Parser</span>

<span class="c1"># first get the list of UID&#39;s</span>
<span class="n">resp</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=hiv&amp;retstart=0&amp;retmax=10&amp;rettype=text&amp;tool=biomed3&amp;format=json&#39;</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">resp</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span>
<span class="n">as_json</span> <span class="o">=</span> <span class="n">resp</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
<span class="n">idlist</span> <span class="o">=</span> <span class="n">as_json</span><span class="p">[</span><span class="s1">&#39;esearchresult&#39;</span><span class="p">][</span><span class="s1">&#39;idlist&#39;</span><span class="p">]</span>

<span class="c1"># now get the data, parse it and cast the content into a list of genomes</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">Parser</span><span class="p">()</span>
<span class="n">genomes</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">idlist</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="using-custom-parsers">
<h2>Using custom parsers<a class="headerlink" href="#using-custom-parsers" title="Permalink to this headline">¶</a></h2>
<p>GenBankParser allows to easily add new and overwrite parsers for specific sections. Here is how you might overwrite the parser for the <code class="docutils literal notranslate"><span class="pre">COMMENT</span></code> section:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">form</span> <span class="n">gbparse</span> <span class="kn">import</span> <span class="nn">Parser</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">Parser</span><span class="p">()</span>


<span class="c1"># define a new parser for the comment section</span>
<span class="k">def</span> <span class="nf">new_comment_parser</span><span class="p">(</span><span class="n">content_lines</span><span class="p">,</span> <span class="n">genome_content</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract the Annotation part from the COMMENT section and save it </span>
<span class="sd">    as an additional &quot;annotation&quot; section to the genome object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_content</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">content_lines</span><span class="p">)</span>
    <span class="n">_annotation_content</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">content_lines</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;::&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
	    <span class="n">_k</span><span class="p">,</span> <span class="n">_v</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="o">.</span><span class="n">strip</span><span class="p">,</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;::&#39;</span><span class="p">))</span>
	    <span class="n">_annotation_content</span><span class="p">[</span><span class="n">_k</span><span class="p">]</span> <span class="o">=</span> <span class="n">_v</span>
    <span class="c1"># add the annotation section</span>
    <span class="n">genome_content</span><span class="p">[</span><span class="s1">&#39;annotation&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_annotation_content</span>
    <span class="c1"># still save the entire comment</span>
    <span class="n">genome_content</span><span class="p">[</span><span class="s1">&#39;comment&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_content</span>


<span class="c1"># now overwrite the comment parser</span>
<span class="n">p</span><span class="o">.</span><span class="n">content_parser</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
    <span class="p">{</span><span class="s1">&#39;comment&#39;</span><span class="p">:</span> <span class="p">{</span><span class="bp">None</span><span class="p">:</span> <span class="n">new_comment_parser</span><span class="p">}}</span>
    <span class="p">)</span>
    
<span class="c1"># DONE! Now, when the parser encounters a COMMENT section,</span>
<span class="c1"># the new_comment_parser method will handle it.</span>
   
</pre></div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">GenBankParser</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">GenBankParser</a></li>
<li class="toctree-l1"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="#package-info">Package info</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#python-compatibility">Python compatibility</a></li>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#accepted-content">Accepted content</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#use-cases">Use cases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#simple-parsing">Simple parsing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#processing">Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#fetch-from-ncbi">Fetch from ncbi</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-custom-parsers">Using custom parsers</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to GenBankParser’s documentation!</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Jonas I. Liechti.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.0.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/README.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>