
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>&lt;no title&gt; &#8212; GenBankParser 0.1.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <p># GenBankParser
__*Unofficial*__ parser for ncbi GenBank data in the _GenBank flatfile <a href="#id39"><span class="problematic" id="id40">format_</span></a>.</p>
<p># Installation
_I recommend using a [virtualenv](<a class="reference external" href="https://pypi.org/project/virtualenv/)!_">https://pypi.org/project/virtualenv/)!_</a></p>
<p>The packages can be pip-installed</p>
<p><code class="docutils literal notranslate"><span class="pre">`bash</span>
<span class="pre">pip</span> <span class="pre">install</span> <span class="pre">git+git://github.com/j-i-l/GenBankParser.git&#64;v0.1.1-alpha</span>
<span class="pre">`</span></code>
<cite>v0.1.1-alpha</cite> is the last version at the moment of writing these instructions.
Check the [releases](<a class="reference external" href="https://github.com/j-i-l/GenBankParser/releases">https://github.com/j-i-l/GenBankParser/releases</a>) section for newer versions.</p>
<p># Package info
<strong>Contributions are welcome!</strong></p>
<p>This packages is not actively maintained.</p>
<p>## Python compatibility
Supported is any python version &gt;=2.7 including python 3.x.</p>
<p>## Requirements
- [configparser](<a class="reference external" href="https://docs.python.org/2/library/configparser.html">https://docs.python.org/2/library/configparser.html</a>)</p>
<blockquote>
<div><p>&gt;pip install configparser</p>
</div></blockquote>
<ul>
<li><p>[requests](<a class="reference external" href="http://docs.python-requests.org/en/master/">http://docs.python-requests.org/en/master/</a>)</p>
<blockquote>
<div><p>&gt;pip install requests</p>
</div></blockquote>
</li>
</ul>
<p>## Accepted content
This GenBankParser aims to parse uncompressed GenBank files in the _GenBank
flatfile <a href="#id41"><span class="problematic" id="id42">format_</span></a>.</p>
<p>They are usually of a form similar to this:</p>
<p><a href="#id1"><span class="problematic" id="id2">``</span></a>`
LOCUS       XXXX             11111111 bp    DNA     circular BCT 01-JAN-2018
DEFINITION  Completely made up, complete genome.
ACCESSION   XXXX
VERSION     XXXX.1  GI:1111111111
DBLINK      BioProject: PRJNA111111</p>
<blockquote>
<div><p>BioSample: SAMN111111</p>
</div></blockquote>
<p>KEYWORDS    .
SOURCE      Completely made up</p>
<blockquote>
<div><dl class="simple">
<dt>ORGANISM  Completely made up</dt><dd><p>Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacterales.</p>
</dd>
</dl>
</div></blockquote>
<p>…</p>
<p><a href="#id3"><span class="problematic" id="id4">``</span></a><a href="#id5"><span class="problematic" id="id6">`</span></a></p>
<p>Accepted are either files with single genomes or genes like [this file](<a class="reference external" href="https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?tool=portal&amp;save=file&amp;log$=seqview&amp;db=nuccore&amp;report=gbwithparts&amp;id=22222&amp;withparts=on">https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?tool=portal&amp;save=file&amp;log$=seqview&amp;db=nuccore&amp;report=gbwithparts&amp;id=22222&amp;withparts=on</a>) or a complete sequence of genomes available from the [NIH genetic sequence database](<a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/">https://www.ncbi.nlm.nih.gov/genbank/</a>).</p>
<p>If you want to process sequence of genomes downloaded from the ncbi GenBank ftp server (<a class="reference external" href="ftp://ftp.ncbi.nih.gov/genbank/">ftp://ftp.ncbi.nih.gov/genbank/</a>), please make sure to first decompress the files before using the GenBankParser.</p>
<p>In addition to GenBank files the GenBankParser also accepts GenBank UIDs or chromosome Genbank identifiers.
GenBankParser then tries to fetch the entries directly from the ncbi database. For an example see the [example](#fetch-from-ncbi) below.</p>
<p>—
# Use cases</p>
<p>## Simple parsing</p>
<p>### Get a list of genomes</p>
<p><a href="#id7"><span class="problematic" id="id8">``</span></a><a href="#id9"><span class="problematic" id="id10">`</span></a>python
from gbparse import Parser</p>
<p>p = Parser()</p>
<p>genome_file = ‘/path/to/genome_file.txt’</p>
<dl class="simple">
<dt>with open(genome_file, ‘r’) as fobj:</dt><dd><p>genomes = p.parse(fobj)</p>
</dd>
</dl>
<p><a href="#id11"><span class="problematic" id="id12">``</span></a><a href="#id13"><span class="problematic" id="id14">`</span></a></p>
<p>### Save genomes as json files to a directory</p>
<p><a href="#id15"><span class="problematic" id="id16">``</span></a><a href="#id17"><span class="problematic" id="id18">`</span></a>python
from gbparse import Parser</p>
<p>p = Parser()</p>
<p>genome_file = ‘/path/to/genome_file.txt’
genomes_save_path = ‘/path/to/genomes/’</p>
<dl class="simple">
<dt>with open(genome_file, ‘r’) as fobj:</dt><dd><p>genomes = p.parse(fobj, genomes_save_path)</p>
</dd>
</dl>
<p><a href="#id19"><span class="problematic" id="id20">``</span></a><a href="#id21"><span class="problematic" id="id22">`</span></a></p>
<p>## Processing</p>
<p>### retrieve set of all present genes in genomes
You might pass a callable to the parser method. The callable needs to accept
a genome (a dictionary) as first argument but can de arbitrary otherwise.
Additional arguments can directly be passed to the parser method.</p>
<p>A simple use-case of a callable would be a method extracting certain
information from each parsed genome, like the set of present genes:</p>
<p><a href="#id23"><span class="problematic" id="id24">``</span></a><a href="#id25"><span class="problematic" id="id26">`</span></a>python
from gbparse import Parser</p>
<p># define a callable that retrieves all genes from a genomes
def get_genes(genome, present_genomes):</p>
<blockquote>
<div><dl>
<dt>present_genomes.extend(</dt><dd><dl>
<dt>list(set(</dt><dd><blockquote>
<div><p>gene.get(‘gene’, None)
for gene in genome[‘content’].get(‘genes’, {})</p>
</div></blockquote>
<p>))</p>
</dd>
</dl>
</dd>
</dl>
<p>)
return None</p>
</div></blockquote>
<p>p = Parser()
# define result variable
list_of_present_genes = []</p>
<p>genome_file = ‘/path/to/genome_file.txt’</p>
<dl class="simple">
<dt>with open(genome_file, ‘r’) as fobj:</dt><dd><p>p.parse(fobj, fct=get_genes, present_genomes=list_of_present_genes)</p>
</dd>
</dl>
<p><a href="#id27"><span class="problematic" id="id28">``</span></a>`
## Fetch from ncbi
Say we want the get the first 10 GenBank files that are returned when searching for ‘hiv’ on the Pubmed database.
Using the [ncbi entrez eutils](<a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25500/">https://www.ncbi.nlm.nih.gov/books/NBK25500/</a>) tool the query to retrieve UID’s of these entries might look like this:</p>
<p>&gt; <a class="reference external" href="https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=hiv&amp;retstart=0&amp;retmax=10&amp;rettype=text&amp;tool=biomed3&amp;format=json">https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=hiv&amp;retstart=0&amp;retmax=10&amp;rettype=text&amp;tool=biomed3&amp;format=json</a></p>
<p>Here is how this can all be done in python:
<a href="#id29"><span class="problematic" id="id30">``</span></a><a href="#id31"><span class="problematic" id="id32">`</span></a>python
import requests
from gbparse import Parser</p>
<p># first get the list of UID’s
resp = requests.get(‘<a class="reference external" href="https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=hiv&amp;retstart=0&amp;retmax=10&amp;rettype=text&amp;tool=biomed3&amp;format=json">https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&amp;term=hiv&amp;retstart=0&amp;retmax=10&amp;rettype=text&amp;tool=biomed3&amp;format=json</a>’)
assert resp.status_code == 200
as_json = resp.json()
idlist = as_json[‘esearchresult’][‘idlist’]</p>
<p># now get the data, parse it and cast the content into a list of genomes
p = Parser()
genomes = p.fetch(idlist)
<a href="#id33"><span class="problematic" id="id34">``</span></a>`
## Using custom parsers
GenBankParser allows to easily add new and overwrite parsers for specific sections. Here is how you might overwrite the parser for the <cite>COMMENT</cite> section:</p>
<p><a href="#id35"><span class="problematic" id="id36">``</span></a><a href="#id37"><span class="problematic" id="id38">`</span></a>python
form gbparse import Parser</p>
<p>p = Parser()</p>
<p># define a new parser for the comment section
def new_comment_parser(content_lines, genome_content):</p>
<blockquote>
<div><p>“””
Extract the Annotation part from the COMMENT section and save it
as an additional “annotation” section to the genome object.
“””
_content = ‘’.join(content_lines)
_annotation_content = {}
for line in content_lines:</p>
<blockquote>
<div><dl class="simple">
<dt>if ‘::’ in line:</dt><dd><p>_k, _v = map(str.strip, line.split(‘::’))
_annotation_content[_k] = _v</p>
</dd>
</dl>
</div></blockquote>
<p># add the annotation section
genome_content[‘annotation’] = _annotation_content
# still save the entire comment
genome_content[‘comment’] = _content</p>
</div></blockquote>
<p># now overwrite the comment parser
p.content_parser.update(</p>
<blockquote>
<div><p>{‘comment’: {None: new_comment_parser}}
)</p>
</div></blockquote>
<p># DONE! Now, when the parser encounters a COMMENT section,
# the new_comment_parser method will handle it.</p>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">GenBankParser</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">GenBankParser</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#package-info">Package info</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html#use-cases">Use cases</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Jonas I. Liechti.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.0.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/readme_link.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>